apps/backend/src/reconciliation/pos-reconciliation.service.ts

```diff
@@ -19,58 +21,83 @@ export class POSReconciliationService {
+    payments: PaymentEntity[],
+    deposits: POSDepositEntity[]
+  ): PosPaymentPosDepositPair[] {
+    const matches: PosPaymentPosDepositPair[] = [];
 
-  private async match(
-    deposits: POSDepositEntity[],
-    payments: PaymentEntity[]
-  ): Promise<any[]> {
-    const filtered = deposits.map((deposit: POSDepositEntity) => ({
-      payment: _.findWhere(payments, {
-        amount: parseFloat(deposit.transaction_amt.toString()),
-        method: deposit.method?.toString()
-      }),
-      deposit
-    }));
+    // Basic EAGER! 1:1 matching
 
-    return (
-      filtered &&
-      (await Promise.all(
-        filtered.map(({ deposit, payment }: any) => {
-          return this.setMatched(deposit, payment);
-        })
-      ))
-    );
+    for (const payment of payments) {
+      // console.log(`Processing payment for ${payment.amount} - ${payment.id} - ${payment.timestamp}`)
+      for (const deposit of deposits) {
+        // console.log(`Processing deposit ${deposit.id} for ${deposit.transaction_amt} - ${deposit.timestamp}`)
+
+        // TODO:make this a strategy pattern
+        if (
+          payment.amount === deposit.transaction_amt &&
+          payment.method === deposit.card_vendor &&
+          differenceInSeconds(payment.timestamp, deposit.timestamp) < 240
+        ) {
+          matches.push({
+            payment,
+            deposit
+          });
+          break;
+        }
+      }
+    }
+
+    for (const match of matches) {
+      match.payment.match = true;
+      match.deposit.match = true;
+      match.payment.deposit_id = match.deposit.id;
+      match.deposit.matched_payment_id = match.payment.id;
+    }
+    return matches;
   }
 
   public async reconcile(
     event: ReconciliationEvent
-  ): Promise<ReconciliationEventOutput | ReconciliationEventError> {
-    const payments = await this.transactionService.queryPosPayments(event);
+  ): Promise<unknown | ReconciliationEventOutput | ReconciliationEventError> {
+    const posPayments = await this.transactionService.queryPosPayments(event);
 
-    const deposits = await this.posDepositService.findAllByLocationAndDate(
-      event
+    if (checkPaymentsForFullMatch(posPayments)) {
+      console.log(`****All payments are already matched for this event`);
+      return;
+    }
+    const alreadyMatchedPosPayments = posPayments.filter((itm) => itm.match);
+    if (posPayments.length === 0) {
+    }
+    const yetToBeMatchedPosPayments = posPayments.filter((itm) => !itm.match);
+
+    const deposits = await this.posDepositService.findPOSDeposits(event);
+    const alreadyMatchedPosDeposits = deposits.filter((itm) => itm.match);
+    const yetToBeMatchedPosDeposits = deposits.filter((itm) => !itm.match);
+
+    const matched_in_this_run = this.matchPosPaymentToPosDeposits(
+      yetToBeMatchedPosPayments,
+      yetToBeMatchedPosDeposits
     );
 
-    const matched =
-      payments && deposits && (await this.match(deposits, payments));
```

Comment:
This issue of having to validate the variable before using it won't occur if we use proper types on all parameters.

Reply:
here payments and deposits are arrays, so I am not sure why we need to validate them, and if they are empty, match function should still be able to handle it,.